### ISSUE TITLE
Engine state no longer keeps a seed?

### ISSUE THREAD

**Issue by achigeor** [pytorch/ignite#1143]
## ❓ Questions/Help/Support

Hi @vfdev-5 

Not sure if that's the best place but I was investing a reproducability issue on my project and found the following:
It seems that in ignite 0.4.0, the engine state no longer holds a `seed` parameter. 
The following is:
* The first one is the val engine saved with `0.4r0.post1`
* The second one is the val engine saved with `0.3.0`
```
OrderedDict([('epoch_length', 504), ('max_epochs', 1), ('iteration', 504)])
OrderedDict([('seed', 12), ('epoch_length', 504), ('max_epochs', 1), ('iteration', 504)])
```
Is this a bug or known?

---- 
I think this might be connected to the problem of reproducability that I'm facing, which is the following. With everything static and same (code, machine, params, torch version, cuda version), running the validation loop of a trained model:
* When using `ignite==0.4r0.post1`, every consecutive run yields different validation score with significant difference (~5%) 
*  When using `ignite==0.3.0`, every run yields the exact same result without any variation at all

## Environment

 - PyTorch Version: 1.3.1
 - Ignite Version: `0.4r0.post1` vs `0.3.0`
 - OS: Linux
 - How you installed Ignite: pip install (for both)
 - Python version: 3.6
 - Any other relevant information: 
    -   GPU: Driver Version: 430.34    CUDA Version: 10.1  
    -  `model.eval()` is properly used

**Comment by vfdev-5** [comment]
@achigeor this is a right place to ask :)

Yes, this is an intended BC breaking change due to various problems we encountered with v0.3.0 and its "deterministic" engine's implementation (see #940 #939 #938).

In few words, now we removed dataflow synchronization and left up to the user to setup manual seed (as internal seed setup is very critical for Bayesian DL and similar).
New Engine normally should produce "almost" the same runs (up to cudnn.benchmark=True etc) if we use it 
like that
```python
from ignite.utils import manual_seed

# handle random init model's params

manual_seed(15)
trainer.run(train_loader, max_epochs=epochs)
```

We added `DeterministicEngine` in #939 in order to reproduce in some sort a previous behaviour.

Complete example of saving and resuming (and reruning) in a "deterministic/reproducible" way can be found here : 
https://github.com/pytorch/ignite/tree/master/examples/mnist#training-save--resume


Let me know what you think about this ?

**Comment by achigeor** [comment]
@vfdev-5  thanks for the quick reply!

I had completely missed this and was expecting the same behaviour. Thanks for pointing it out! 

Just tried with `manual_seed()` and has the behaviour I was expecting and the reproducability tolerance we look for.

I also got a question regarding `DeterministicEngine`. What is the trade off of using this instead of `Engine`? I mean what is the use case of _not_ using it?

**Comment by vfdev-5** [comment]
> I also got a question regarding DeterministicEngine. What is the trade off of using this instead of Engine? I mean what is the use case of not using it?

Main trade off is the training speed due to 

> Internally, `torch.backends.cudnn.deterministic = True` and `torch.backends.cudnn.benchmark = False` are also applied.

in [DeterministicEngine](https://pytorch.org/ignite/engine.html#ignite.engine.deterministic.DeterministicEngine). 

For other unintended behaviours that could happens with `DeterministicEngine`, please see all "Warnings" in https://pytorch.org/ignite/concepts.html#dataflow-synchronization.

But on the other hand, in *certain configuration* we can have the same gradients with `DeterministicEngine`...

**Comment by achigeor** [comment]
I see makes sense. Thank you for the clarifications @vfdev-5! I will also pay more attention to the release logs next time, just saw the above were listed in the release :)
